## Day 49 Lecture 2 Assignment

In this assignment, we will apply mean shift clustering to a dataset containing the results of a survey on financial wellbeing.

In [ ]:
%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.cluster import MeanShift
from sklearn.preprocessing import StandardScaler

This dataset contains the results of a survey on a financial wellbeing conducted by the US Consumer Finance Protection Bureau that were published in October 2017. This dataset has a large number of columns, most of which correspond to specific questions on the survey. These codebook for translating the column names to questions can be found here:

https://s3.amazonaws.com/files.consumerfinance.gov/f/documents/cfpb_nfwbs-puf-codebook.pdf

Load the dataset.

In [ ]:
# answer goes here
df = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/Data%20Sets%20Clustering/financial_wellbeing.csv')

df


PUF_ID  sample  fpl  SWB_1  ...  PPT1317  PPT18OV  PCTLT200FPL   finalwt
0      10350       2    3      5  ...        0        1            0  0.367292
1       7740       1    3      6  ...        0        2            0  1.327561
2      13699       1    3      4  ...        1        2            1  0.835156
3       7267       1    3      6  ...        0        1            0  1.410871
4       7375       1    3      4  ...        0        4            1  4.260668
...      ...     ...  ...    ...  ...      ...      ...          ...       ...
6389   11220       3    3      6  ...        1        2           -5  0.522504
6390   13118       3    2      7  ...        0        3           -5  1.015219
6391    8709       1    3      5  ...        0        2            0  1.136270
6392    8515       1    3      5  ...        0        2            0  1.224941
6393    8516       3    2      4  ...        0        2           -5  0.626489

[6394 rows x 217 columns]

While the survey questions have the potential for interesting cluster analysis, we will stick to the "score" columns to avoid clustering in an unreasonably high-dimensional space. The columns we are interested in all have "score" in their names; identify and isolate these columns. (There should be 4 in total.)

In [ ]:
# answer goes here
cols = [col for col in df.columns if 'score' in col]
print(cols)

['FWBscore', 'FSscore', 'LMscore', 'KHscore']


In [ ]:
X = pd.concat([df['FWBscore'],df['FSscore'],df['LMscore'],df['KHscore']], axis=1)
X

FWBscore  FSscore  LMscore  KHscore
0           55       44        3    1.267
1           51       43        3   -0.570
2           49       42        3   -0.188
3           49       42        2   -1.485
4           49       42        1   -1.900
...        ...      ...      ...      ...
6389        61       47        3    1.267
6390        59       59        1   -1.215
6391        59       51        2   -1.215
6392        46       54        2   -1.215
6393        49       42        2   -1.215

[6394 rows x 4 columns]

Standardize the features in your dataset using scikit-learn's StandardScaler, which will set the mean of each feature to 0 and the variance to 1.

In [ ]:
# answer goes here
scale = StandardScaler()
X_scale = pd.DataFrame(scale.fit_transform(X), columns=X.columns)
X_scale




FWBscore   FSscore   LMscore   KHscore
0    -0.073062 -0.530906  0.653830  1.624716
1    -0.355677 -0.609920  0.653830 -0.629626
2    -0.496984 -0.688935  0.653830 -0.160841
3    -0.496984 -0.688935 -0.670399 -1.752502
4    -0.496984 -0.688935 -1.994628 -2.261785
...        ...       ...       ...       ...
6389  0.350859 -0.293863  0.653830  1.624716
6390  0.209552  0.654309 -1.994628 -1.421162
6391  0.209552  0.022194 -0.670399 -1.421162
6392 -0.708944  0.259237 -0.670399 -1.421162
6393 -0.496984 -0.688935 -0.670399 -1.421162

[6394 rows x 4 columns]

Run mean shift clustering on the scores in the survey dataset using the default bandwidth. Then answer the following by printing or typing as appropriate:

- How many clusters are produced? 
- What are the cluster centers?
- How many responses are assigned to each cluster?
- Are these results reasonable? If not, what changes should we make?

In [ ]:
# answer goes here
ms = MeanShift()
X['ms_cluster'] = ms.fit_predict(X_scale)
X['ms_cluster'].value_counts()




0    6326
1      68
Name: ms_cluster, dtype: int64

In [ ]:
ms.cluster_centers_

array([[ 0.09918758, -0.04722472,  0.40797515,  0.33667793],
       [-4.02966165, -4.08655136, -2.98779955, -2.17097301]])



*   2 clusters
*   [ 0.09918758, -0.04722472,  0.40797515,  0.33667793],  [-4.02966165, -4.08655136, -2.98779955, -2.17097301]
*   cluster 0: 6326, cluster 1: 68
*   The result is not reasonable. We can try adjusting bandwidth to make the clusters more evenly spread.





Try changing the appropriate parameters of the mean shift algorithm to achieve a better clustering result. Answer all of the same questions from the previous clustering step.

In [ ]:
# answer goes here
ms = MeanShift(bandwidth=1.32, bin_seeding=True)
X['ms_cluster'] = ms.fit_predict(X_scale)

print('16 clusters and responses in each is:')
X['ms_cluster'].value_counts()

16 clusters and responses in each is:


0     3774
1     1719
2      492
3       96
4       85
5       45
12      40
15      28
7       27
13      23
9       23
14      17
11      12
8        6
6        5
10       2
Name: ms_cluster, dtype: int64

In [ ]:
print('centers are: \n',ms.cluster_centers_)

centers are: 
 [[ 0.18203951 -0.00987377  0.65383032  0.59416646]
 [-0.25701796 -0.33807842 -0.67039872 -0.66450272]
 [-0.48160626 -0.57506123 -1.99462776 -1.22059989]
 [-0.55767339 -0.39617658 -3.31885681 -1.5359196 ]
 [ 0.09650608  1.78421394 -3.31885681 -1.53455396]
 [-1.38015326  2.25434925 -3.31885681 -1.50399701]
 [-4.02966165 -4.08655136 -2.98779955 -2.17097301]
 [ 1.24580395 -2.90133629  0.21242064 -1.47147656]
 [ 0.51571719 -3.79683212 -3.31885681 -2.4495444 ]
 [ 2.75307983  2.70868169 -3.31885681 -2.4495444 ]
 [ 1.69327648 -3.21739364  0.65383032  0.94362696]
 [ 0.35085889 -3.37542232 -1.99462776 -1.75250233]
 [ 0.20955178 -2.74330761 -0.67039872  1.62471561]
 [-1.27417292 -2.90133629 -3.31885681 -0.62962645]
 [-2.19266917 -0.68893483 -3.31885681  0.36684919]
 [-4.24162232 -0.21484881 -1.99462776 -0.62962645]]


*   This result is more reasonable. I think the main issue here is that this dataset doesn't fit meanshift well. 



